### PR TITLE
Add integration test with infrastructure

### PR DESCRIPTION
**Pull request by folex**
- Migrate from Rust 2015 to Rust 2018!
- Add `tests/` with `pinning.rs` and `utils/mod.rs`
- Add `lib.rs` to make integration tests possible
- Move all crates & import to `lib.rs`
- Remove most of `extern crate`s, only one left (there is a comment on why)
- Add a few CLI usage examples
- Remove accidental doc-tests in `src/utils.rs`

### CODE REVIEW

File: cli/README.md [hunk 1]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
 
 To look at all possible arguments and options use `./fluence --help`:
 
+You can use `./fluence [SUBCOMMAND] --help` to learn how to use commands.
+
+## Usage examples
+### Register a node (usually happens automatically via compose.sh)
```
Comment (DieMyst) [inline cli/README.md]: info in brackets is excess

File: cli/README.md [hunk 2]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+The following command will register a node with the following parameters:
```
Comment (DieMyst) [inline cli/README.md]: some words about what is `register` and where it will be registered are necessary

File: cli/README.md [hunk 3]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+- advertised address `127.0.0.1`
```
Comment (DieMyst) [inline cli/README.md]: `127.0.0.1` is only for test purposes, some words about external IP is necessary. And what will be done through this IP address?

File: cli/README.md [hunk 4]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+- Tendermint key (also used to identify node) `1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM=`, note base64 format
```
Comment (DieMyst) [inline cli/README.md]: Some words about what contains in tendermint key.
Btw, if we pass tendermint key as an argument, we can get info about IP address from it. Maybe add TODO about this?

File: cli/README.md [hunk 5]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+    - flag `--base64_tendermint_key` passed so tendermint key is treated as base64-encoded as opposed to hex-encoded
+-  `0x4180fc65d613ba7e1a385181a219f1dbfe7bf11d` will be used as Ethereum account for issuing transactions
```
Comment (DieMyst) [inline cli/README.md]: if this text describes an only local test usage, I think, you need to describe this in text and describe where did these arguments come from.

File: cli/README.md [hunk 6]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
-    register            Register worker in smart contract
-    status              Get status of smart contract
+./fluence register 127.0.0.1 1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM= 0x4180fc65d613ba7e1a385181a219f1dbfe7bf11d 0x9995882876ae612bfd829498ccd73dd962ec950a --base64_tendermint_key --secret-key 0xcb0799337df06a6c73881bab91304a68199a430ccd4bc378e37e51fd1b118133 --wait_syncing --start_port 25000 --last_port 25010
 ```
 
-You can use `./fluence [SUBCOMMAND] --help` to learn how to use commands.
+### Publish app
+The following command will publish app `counter.wasm`. Interesting bits:
+- `--cluster_size 4` requires cluster of 4 workers to host this app
+- `--pin_to 1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM= --base64` requires that one of the node in cluster must be `1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM=`
```
Comment (DieMyst) [inline cli/README.md]: and this node should be registered before this operation

File: cli/README.md [hunk 7]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+```
+./fluence publish fluence/vm/examples/counter/target/wasm32-unknown-unknown/release/deps/counter.wasm 0x9995882876ae612bfd829498ccd73dd962ec950a 0x4180fc65d613ba7e1a385181a219f1dbfe7bf11d --cluster_size 4 --secret-key 0xcb0799337df06a6c73881bab91304a68199a430ccd4bc378e37e51fd1b118133 --pin_to 1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM= --base64
```
Comment (DieMyst) [inline cli/README.md]: we don't have this file if I remember properly

File: cli/README.md [hunk 8]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
 
 To look at all possible arguments and options use `./fluence --help`:
 
+You can use `./fluence [SUBCOMMAND] --help` to learn how to use commands.
+
+## Usage examples
```
Comment (DieMyst) [inline cli/README.md]: A few words about what you need to have eth to use this CLI

File: cli/README.md [hunk 9]
```diff
@@ -26,26 +26,35 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
-    check               Verifies wasm file, issue warning for using functions from banned modules.
-    help                Prints this message or the help of the given subcommand(s)
-    publish             Publish code to ethereum blockchain
-    register            Register worker in smart contract
-    status              Get status of smart contract
+./fluence register 127.0.0.1 1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM= 0x4180fc65d613ba7e1a385181a219f1dbfe7bf11d 0x9995882876ae612bfd829498ccd73dd962ec950a --base64_tendermint_key --secret-key 0xcb0799337df06a6c73881bab91304a68199a430ccd4bc378e37e51fd1b118133 --wait_syncing --start_port 25000 --last_port 25010
 ```
 
-You can use `./fluence [SUBCOMMAND] --help` to learn how to use commands.
+### Publish app
```
Comment (DieMyst) [inline cli/README.md]: describe what does it mean

File: cli/README.md (hunk 4 above)
Comment (folex) [reply cli/README.md]: Can you please share a few thoughts about content of Tendermint key?

File: cli/README.md [hunk 10]
```diff
@@ -26,26 +24,132 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+Parameters are:
+- advertised address `85.82.118.4`, please note that this address should be available from Internet as it will be used to connect with other workers in a future cluster
+- Tendermint key (used to identify node) `1GVDICzgrw1qahPfSbwCfYw0zrw91OMZ46QoKvJMjjM=`, note base64 format
+    - flag `--base64_tendermint_key` passed so tendermint key is treated as base64-encoded as opposed to hex-encoded
+    - currently, Tendermint key can be found in logs of `fluencelabs/node` Docker container
+    - note that key should be unique, i.e. you can't register several nodes with the same key
+- `0x4180fc65d613ba7e1a385181a219f1dbfe7bf11d` will be used as Ethereum account for issuing transactions. _Use your Ethereum account here_
+- `0x9995882876ae612bfd829498ccd73dd962ec950a` is a contract address, register transaction will be sent there
+- `--secret_key 0xcb0799337df06a6c73881bab91304a68199a430ccd4bc378e37e51fd1b118133` denotes an Ethereum private key, used for offline transaction signing. _Use your Ethereum private key here_
+    - using `--password` is possible instead of private key, but private key is preferred
```
Comment (DieMyst) [inline cli/README.md]: Account must be presented in ethereum node for password usage

File: cli/README.md [hunk 11]
```diff
@@ -26,26 +24,132 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+To deploy your app on Fluence network, you must upload it to Swarm and publish hash to smart-contract. The simplest way to achieve that is to use CLI command `publish`.
 
-SUBCOMMANDS:
-    add-to-whitelist    Adds an address to the whitelist of Fluence smart contract
-    check               Verifies wasm file, issue warning for using functions from banned modules.
-    help                Prints this message or the help of the given subcommand(s)
-    publish             Publish code to ethereum blockchain
-    register            Register worker in smart contract
-    status              Get status of smart contract
+The following command will publish app `counter.wasm`.
```
Comment (DieMyst) [inline cli/README.md]: You can publish only wasm apps, link to rust sdk for example

File: cli/README.md [hunk 12]
```diff
@@ -26,26 +24,132 @@ Requirements: [`cargo`](https://doc.rust-lang.org/cargo/getting-started/installa
...
+```
+./fluence publish \
+            fluence/vm/examples/counter/target/wasm32-unknown-unknown/release/deps/counter.wasm \
```
Comment (DieMyst) [inline cli/README.md]: Use some `path/to/file/counter.wasm` to abstract from real paths

Comment (DieMyst) [reply cli/README.md]: Or move info about this paths and counter.wasm from 'interesting bits' here

File: cli/tests/utils/mod.rs [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    gas: u32,
+    code_bytes: Vec<u8>,
+    swarm_url: String,
+}
+
+impl TestOpts {
+    pub fn default() -> TestOpts {
+        TestOpts {
+            contract_address: "9995882876ae612bfd829498ccd73dd962ec950a".parse().unwrap(),
+            account: "4180fc65d613ba7e1a385181a219f1dbfe7bf11d".parse().unwrap(),
```
Comment (DieMyst) [inline cli/tests/utils/mod.rs]: Comment that you use account and addresses from ganache.

File: cli/tests/pinning.rs [hunk 14]
```diff
@@ -0,0 +1,35 @@
+mod utils;
+
+use crate::utils::*;
+use fluence::contract_func::contract::events::app_deployed;
+use fluence::contract_status::status::{get_status, Status};
+use fluence::register::Register;
+use web3::types::H256;
+
+#[test]
+fn integration_publish_pinned() -> () {
```
Comment (DieMyst) [inline cli/tests/pinning.rs]: Maybe it is possible to start ganache not in Travis, but from the rust code

**Pull request merged**